File: src/main/java/com/github/lavenderx/netty_rpc/protocol/RpcEncoder.java
package com.github.lavenderx.netty_rpc.protocol;

import com.github.lavenderx.netty_rpc.utils.SerializationUtils;
import io.netty.buffer.ByteBuf;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.codec.MessageToByteEncoder;

public class RpcEncoder extends MessageToByteEncoder {

    private final Class<?> genericClass;

    public RpcEncoder(Class<?> genericClass) {
        this.genericClass = genericClass;
    }

    @Override
    public void encode(ChannelHandlerContext ctx, Object in, ByteBuf out) throws Exception {
        if (genericClass.isInstance(in)) {
            byte[] data = SerializationUtils.serialize(in);
            out.writeInt(data.length);
            out.writeBytes(data);
        }
    }
}

File: foliage-rpc/src/main/kotlin/org/lavenderx/foliage/nettyrpc/protocol/RpcEncoder.kt
package org.lavenderx.foliage.nettyrpc.protocol

import io.netty.buffer.ByteBuf
import io.netty.channel.ChannelHandlerContext
import io.netty.handler.codec.MessageToByteEncoder
import org.lavenderx.foliage.nettyrpc.utils.SerializationUtils

class RpcEncoder<T : Any>(private val genericClass: Class<T>) : MessageToByteEncoder<Any>() {

    @Throws(Exception::class)
    override fun encode(ctx: ChannelHandlerContext, msg: Any, out: ByteBuf) {
        if (genericClass.isInstance(msg)) {
            val data = SerializationUtils.serialize(msg)
            out.writeInt(data.size)
            out.writeBytes(data)
        }
    }
}